streams/src/test/java/org/apache/kafka/streams/StreamsBuilderTest.java

```diff
@@ -598,7 +599,7 @@ public void shouldUseSpecifiedNameForLeftJoinOperationBetweenKStreamAndKStream()
         final KStream<String, String> streamOne = builder.stream(STREAM_TOPIC);
         final KStream<String, String> streamTwo = builder.stream(STREAM_TOPIC_TWO);
 
-        streamOne.leftJoin(streamTwo, (value1, value2) -> value1, JoinWindows.of(Duration.ofHours(1)), Joined.as(STREAM_OPERATION_NAME));
+        streamOne.leftJoin(streamTwo, (value1, value2) -> value1, JoinWindows.of(Duration.ofHours(1)), StreamJoined.<String, String, String>as(STREAM_OPERATION_NAME).withName(STREAM_OPERATION_NAME));
```

Comment:
Can you elaborate on backward compatibility? Why do we need to update the test? Does this change indicate an incompatibility?

Reply:
This change to the test gets to the point of KIP-479.  Previously by naming the Join operator via `Joined` we were naming the processor, repartition topic, **_and_** the state store (hence changelog topic).  

But that is a breaking change for users using `Joined` going from  2.3 to 2.4.  So instead of changing the expected topology, IMHO it's better to use `StreamJoined` to explicitly name the statestore here.  CF my previous comments https://github.com/apache/kafka/pull/7285#discussion_r320792207

Reply:
So the change is just to use the new API. If we would leave `Joined.as()` the test would still pass.

Should we have this test for old and new API to test compatibility? Or this this covered elsewhere?

Reply:
If we kept `Joined.as()` the test will fail as it would not name the state store so the generated topology would be different from the expected topology.
But this is not a breaking change as only `trunk` has the behavior of Joined.as() naming the state stores.

I can add a separate test using joined with a different topology showing that the store name is generated now.
WDYT?

Reply:
Ack. Thanks for the details.